Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Manager.swift

```diff
@@ -12,7 +12,157 @@
 //
 //===----------------------------------------------------------------------===//
 
+import Logging
+import NIO
 import NIOConcurrencyHelpers
+import NIOHTTP1
+
+extension HTTPConnectionPool {
+    final class Manager {
+        private typealias Key = ConnectionPool.Key
+
+        enum State {
+            case active
+            case shuttingDown(promise: EventLoopPromise<Bool>, unclean: Bool)
+            case shutDown
+        }
+
+        let eventLoopGroup: EventLoopGroup
+        let configuration: HTTPClient.Configuration
+        let connectionIDGenerator = Connection.ID.globalGenerator
+        let logger: Logger
+
+        private var state: State = .active
+        private var _pools: [Key: HTTPConnectionPool] = [:]
+        private let lock = Lock()
+
+        private let sslContextCache = SSLContextCache()
+
+        init(eventLoopGroup: EventLoopGroup,
+             configuration: HTTPClient.Configuration,
+             backgroundActivityLogger logger: Logger) {
+            self.eventLoopGroup = eventLoopGroup
+            self.configuration = configuration
+            self.logger = logger
+        }
+
+        deinit {
+            guard case .shutDown = self.state else {
```

Comment:
What about shutting down? You could call shutdown and not bother waiting for the future to complete as well, I assume?